#include "LoginDialog.h"
#include "ui_LoginDialog.h"
#include "../utils/AdminAuth.h"
#include <QMessageBox>
#include <QTimer>
#include <QApplication>
#include <QLoggingCategory>

Q_LOGGING_CATEGORY(loginDialog, "qkchat.server.admin.logindialog")

LoginDialog::LoginDialog(QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::LoginDialog)
    , _adminAuth(new AdminAuth(this))
    , _lockoutTimer(new QTimer(this))
    , _failedAttempts(0)
    , _isDarkTheme(false)
{
    ui->setupUi(this);
    
    setupUI();
    connectSignals();
    loadSettings();
    
    // Âä†ËΩΩÁÆ°ÁêÜÂëòÈÖçÁΩÆ
    _adminAuth->loadConfig();
    
    qCInfo(loginDialog) << "Login dialog initialized";
}

LoginDialog::~LoginDialog()
{
    delete ui;
}

void LoginDialog::setupUI()
{
    setWindowTitle("QK Chat ÊúçÂä°Âô®ÁÆ°ÁêÜ");
    setFixedSize(400, 300);
    setModal(true);
    
    // ËÆæÁΩÆÁ™óÂè£ÂõæÊ†á
    setWindowIcon(QIcon(":/icons/server.png"));
    
    // ÂàùÂßãÂåñUIÁªÑ‰ª∂
    ui->usernameLineEdit->setPlaceholderText("ÁÆ°ÁêÜÂëòÁî®Êà∑Âêç");
    ui->passwordLineEdit->setPlaceholderText("ÁÆ°ÁêÜÂëòÂØÜÁ†Å");
    ui->passwordLineEdit->setEchoMode(QLineEdit::Password);
    
    // ËÆæÁΩÆÊåâÈíÆ
    ui->loginButton->setText("ÁôªÂΩï");
    ui->loginButton->setDefault(true);
    ui->cancelButton->setText("ÂèñÊ∂à");
    
    // ‰∏ªÈ¢òÂàáÊç¢ÊåâÈíÆ
    ui->themeButton->setText(_isDarkTheme ? "üåû" : "üåô");
    ui->themeButton->setToolTip("ÂàáÊç¢‰∏ªÈ¢ò");
    
    // Áä∂ÊÄÅÊ†áÁ≠æ
    ui->statusLabel->setText("");
    ui->statusLabel->setStyleSheet("color: red;");
    
    // ËÆ∞‰ΩèÂØÜÁ†ÅÂ§çÈÄâÊ°Ü
    ui->rememberCheckBox->setText("ËÆ∞‰ΩèÂØÜÁ†Å");
    
    // ÈîÅÂÆöÁä∂ÊÄÅÊ†áÁ≠æÔºàÂàùÂßãÈöêËóèÔºâ
    ui->lockoutLabel->setVisible(false);
    ui->lockoutLabel->setStyleSheet("color: red; font-weight: bold;");
    
    updateTheme();
}

void LoginDialog::connectSignals()
{
    // ÊåâÈíÆ‰ø°Âè∑
    connect(ui->loginButton, &QPushButton::clicked, this, &LoginDialog::onLoginClicked);
    connect(ui->cancelButton, &QPushButton::clicked, this, &QDialog::reject);
    connect(ui->themeButton, &QPushButton::clicked, this, &LoginDialog::onToggleTheme);
    
    // ÂõûËΩ¶ÈîÆÁôªÂΩï
    connect(ui->usernameLineEdit, &QLineEdit::returnPressed, this, &LoginDialog::onLoginClicked);
    connect(ui->passwordLineEdit, &QLineEdit::returnPressed, this, &LoginDialog::onLoginClicked);
    
    // ÁÆ°ÁêÜÂëòËÆ§ËØÅ‰ø°Âè∑
    connect(_adminAuth, &AdminAuth::accountLocked, this, &LoginDialog::onAccountLocked);
    connect(_adminAuth, &AdminAuth::accountUnlocked, this, &LoginDialog::onAccountUnlocked);
    connect(_adminAuth, &AdminAuth::authenticationFailed, this, &LoginDialog::onAuthenticationFailed);
    
    // ÈîÅÂÆöËÆ°Êó∂Âô®
    _lockoutTimer->setSingleShot(false);
    _lockoutTimer->setInterval(1000); // ÊØèÁßíÊõ¥Êñ∞
    connect(_lockoutTimer, &QTimer::timeout, this, &LoginDialog::updateLockoutStatus);
}

void LoginDialog::loadSettings()
{
    QSettings settings;
    
    // Âä†ËΩΩ‰∏ªÈ¢òËÆæÁΩÆ
    _isDarkTheme = settings.value("admin/dark_theme", false).toBool();
    
    // Âä†ËΩΩËÆ∞‰ΩèÁöÑÁî®Êà∑Âêç
    if (settings.value("admin/remember_password", false).toBool()) {
        ui->rememberCheckBox->setChecked(true);
        ui->usernameLineEdit->setText(settings.value("admin/username", "").toString());
        // Ê≥®ÊÑèÔºöÂØÜÁ†Å‰∏ç‰øùÂ≠òÔºå‰ªÖ‰øùÂ≠òÁî®Êà∑Âêç
    }
    
    // Ê£ÄÊü•ÊòØÂê¶ÊúâÈîÅÂÆöÁä∂ÊÄÅ
    if (_adminAuth->isAccountLocked()) {
        onAccountLocked();
    }
}

void LoginDialog::saveSettings()
{
    QSettings settings;
    
    // ‰øùÂ≠ò‰∏ªÈ¢òËÆæÁΩÆ
    settings.setValue("admin/dark_theme", _isDarkTheme);
    
    // ‰øùÂ≠òËÆ∞‰ΩèÂØÜÁ†ÅËÆæÁΩÆ
    settings.setValue("admin/remember_password", ui->rememberCheckBox->isChecked());
    
    if (ui->rememberCheckBox->isChecked()) {
        settings.setValue("admin/username", ui->usernameLineEdit->text());
    } else {
        settings.remove("admin/username");
    }
}

void LoginDialog::onLoginClicked()
{
    QString username = ui->usernameLineEdit->text().trimmed();
    QString password = ui->passwordLineEdit->text();
    
    // Âü∫Êú¨È™åËØÅ
    if (username.isEmpty() || password.isEmpty()) {
        showStatusMessage("ËØ∑ËæìÂÖ•Áî®Êà∑ÂêçÂíåÂØÜÁ†Å", true);
        return;
    }
    
    // Ê£ÄÊü•Ë¥¶Êà∑ÊòØÂê¶Ë¢´ÈîÅÂÆö
    if (_adminAuth->isAccountLocked()) {
        int remainingTime = _adminAuth->getRemainingLockoutTime();
        showStatusMessage(QString("Ë¥¶Êà∑Â∑≤Ë¢´ÈîÅÂÆöÔºåËØ∑Á≠âÂæÖ %1 Áßí").arg(remainingTime), true);
        return;
    }
    
    // Á¶ÅÁî®ÁôªÂΩïÊåâÈíÆÈò≤Ê≠¢ÈáçÂ§çÁÇπÂáª
    ui->loginButton->setEnabled(false);
    ui->loginButton->setText("ÁôªÂΩï‰∏≠...");
    
    // ÊâßË°åËÆ§ËØÅ
    bool authResult = _adminAuth->authenticate(username, password);
    
    if (authResult) {
        // ËÆ§ËØÅÊàêÂäü
        saveSettings();
        showStatusMessage("ÁôªÂΩïÊàêÂäü", false);
        
        qCInfo(loginDialog) << "Admin login successful:" << username;
        
        // Áü≠ÊöÇÂª∂ËøüÂêéÊé•ÂèóÂØπËØùÊ°Ü
        QTimer::singleShot(500, this, &QDialog::accept);
    } else {
        // ËÆ§ËØÅÂ§±Ë¥•
        _failedAttempts = _adminAuth->getFailedAttempts();
        
        // Ê∏ÖÁ©∫ÂØÜÁ†ÅÂ≠óÊÆµ
        ui->passwordLineEdit->clear();
        ui->passwordLineEdit->setFocus();
        
        // ÈáçÊñ∞ÂêØÁî®ÁôªÂΩïÊåâÈíÆ
        ui->loginButton->setEnabled(true);
        ui->loginButton->setText("ÁôªÂΩï");
        
        qCWarning(loginDialog) << "Admin login failed for user:" << username 
                              << "Attempts:" << _failedAttempts;
    }
}

void LoginDialog::onToggleTheme()
{
    _isDarkTheme = !_isDarkTheme;
    ui->themeButton->setText(_isDarkTheme ? "üåû" : "üåô");
    updateTheme();
}

void LoginDialog::onAccountLocked()
{
    ui->loginButton->setEnabled(false);
    ui->usernameLineEdit->setEnabled(false);
    ui->passwordLineEdit->setEnabled(false);
    
    ui->lockoutLabel->setVisible(true);
    
    // ÂêØÂä®ÂÄíËÆ°Êó∂ËÆ°Êó∂Âô®
    _lockoutTimer->start();
    updateLockoutStatus();
    
    showStatusMessage("Ë¥¶Êà∑Â∑≤Ë¢´ÈîÅÂÆö", true);
    
    qCWarning(loginDialog) << "Account locked due to too many failed attempts";
}

void LoginDialog::onAccountUnlocked()
{
    ui->loginButton->setEnabled(true);
    ui->usernameLineEdit->setEnabled(true);
    ui->passwordLineEdit->setEnabled(true);
    
    ui->lockoutLabel->setVisible(false);
    _lockoutTimer->stop();
    
    _failedAttempts = 0;
    showStatusMessage("Ë¥¶Êà∑Â∑≤Ëß£ÈîÅ", false);
    
    qCInfo(loginDialog) << "Account unlocked";
}

void LoginDialog::onAuthenticationFailed(const QString &reason)
{
    showStatusMessage(reason, true);
    
    // ÊòæÁ§∫Ââ©‰ΩôÂ∞ùËØïÊ¨°Êï∞
    int remainingAttempts = MAX_FAILED_ATTEMPTS - _adminAuth->getFailedAttempts();
    if (remainingAttempts > 0 && !_adminAuth->isAccountLocked()) {
        QString message = QString("Ââ©‰ΩôÂ∞ùËØïÊ¨°Êï∞: %1").arg(remainingAttempts);
        ui->statusLabel->setText(ui->statusLabel->text() + "\n" + message);
    }
}

void LoginDialog::updateLockoutStatus()
{
    if (!_adminAuth->isAccountLocked()) {
        onAccountUnlocked();
        return;
    }
    
    int remainingTime = _adminAuth->getRemainingLockoutTime();
    if (remainingTime <= 0) {
        _adminAuth->unlockAccount();
        return;
    }
    
    int minutes = remainingTime / 60;
    int seconds = remainingTime % 60;
    
    QString lockoutText = QString("Ë¥¶Êà∑ÈîÅÂÆö‰∏≠ÔºåÂâ©‰ΩôÊó∂Èó¥: %1:%2")
                         .arg(minutes, 2, 10, QChar('0'))
                         .arg(seconds, 2, 10, QChar('0'));
    
    ui->lockoutLabel->setText(lockoutText);
}

void LoginDialog::showStatusMessage(const QString &message, bool isError)
{
    ui->statusLabel->setText(message);
    ui->statusLabel->setStyleSheet(isError ? "color: red;" : "color: green;");
    
    // Â¶ÇÊûúÊòØÊàêÂäüÊ∂àÊÅØÔºå3ÁßíÂêéÊ∏ÖÁ©∫
    if (!isError) {
        QTimer::singleShot(3000, this, [this]() {
            ui->statusLabel->clear();
        });
    }
}

void LoginDialog::updateTheme()
{
    QString styleSheet;
    
    if (_isDarkTheme) {
        // Ê∑±Ëâ≤‰∏ªÈ¢ò
        styleSheet = R"(
            QDialog {
                background-color: #2b2b2b;
                color: #ffffff;
            }
            QLineEdit {
                background-color: #3c3c3c;
                border: 1px solid #555555;
                border-radius: 4px;
                padding: 8px;
                color: #ffffff;
            }
            QLineEdit:focus {
                border: 2px solid #0078d4;
            }
            QPushButton {
                background-color: #0078d4;
                border: none;
                border-radius: 4px;
                padding: 8px 16px;
                color: #ffffff;
                font-weight: bold;
            }
            QPushButton:hover {
                background-color: #106ebe;
            }
            QPushButton:pressed {
                background-color: #005a9e;
            }
            QPushButton:disabled {
                background-color: #555555;
                color: #999999;
            }
            QCheckBox {
                color: #ffffff;
            }
            QCheckBox::indicator {
                width: 16px;
                height: 16px;
            }
            QCheckBox::indicator:unchecked {
                background-color: #3c3c3c;
                border: 1px solid #555555;
            }
            QCheckBox::indicator:checked {
                background-color: #0078d4;
                border: 1px solid #0078d4;
            }
        )";
    } else {
        // ÊµÖËâ≤‰∏ªÈ¢ò
        styleSheet = R"(
            QDialog {
                background-color: #ffffff;
                color: #000000;
            }
            QLineEdit {
                background-color: #ffffff;
                border: 1px solid #cccccc;
                border-radius: 4px;
                padding: 8px;
                color: #000000;
            }
            QLineEdit:focus {
                border: 2px solid #0078d4;
            }
            QPushButton {
                background-color: #0078d4;
                border: none;
                border-radius: 4px;
                padding: 8px 16px;
                color: #ffffff;
                font-weight: bold;
            }
            QPushButton:hover {
                background-color: #106ebe;
            }
            QPushButton:pressed {
                background-color: #005a9e;
            }
            QPushButton:disabled {
                background-color: #cccccc;
                color: #666666;
            }
            QCheckBox {
                color: #000000;
            }
            QCheckBox::indicator {
                width: 16px;
                height: 16px;
            }
            QCheckBox::indicator:unchecked {
                background-color: #ffffff;
                border: 1px solid #cccccc;
            }
            QCheckBox::indicator:checked {
                background-color: #0078d4;
                border: 1px solid #0078d4;
            }
        )";
    }
    
    setStyleSheet(styleSheet);
} 